Improve k8s job error handling

ilveroluca opened the pull request:
This PR should fix cases where successful Kubernetes jobs whose logs have gone missing can't be marked as finished by Galaxy. The current implementation trips over an IOError before reaching the code where the job is marked as finished.

In addition, the PR adds logging to DEBUG of full Kubernetes job objects that are in an unexpected state.  The idea is that this debug output should help users of the k8s runner debug an issue that has been with us for a while.

ilveroluca committed:
Catch IOErrors when the k8s runner tries to collect job logs

This commit should fix cases where successful job can't be marked as
done by Galaxy because the logs have gone missing.

ilveroluca committed:
Print full job object to the DEBUG log if k8s runner finds it in an unrecognized state

pcm32 commented on lib/galaxy/jobs/runners/kubernetes.py:
```diff
@@ -515,8 +520,13 @@ def fail_job(self, job_state):
+            with open(job_state.output_file, 'r') as outfile:
+                stdout_content = outfile.read()
+        except IOError as e:
+            stdout_content = "<Failed to recuperate log>"
```
Maybe here indicate the kubernetes job or pod identifier?

pcm32 replied:
I mean for the stdout_content...

ilveroluca replied:
Good thinking @pcm32.  I just pushed [a commit with that addition](https://github.com/ilveroluca/galaxy/commit/cd6c4d4db8bcb72ae931797c216b9829313850cf).  I figure it should appear in this PR as soon as github solves their issues they're experiencing today.

pcm32 reviewed (approved):
LGTM! Thanks @ilveroluca!

ilveroluca committed:
Add job id to missing k8s log text

pcm32 commented:
@bgruening I have reviewed this but we need someone with higher rank than me to re-review/approve this, so if you could tag someone or do yourself it would be great. I know that @ilveroluca has been using it in some of our instances without issues as well.

bgruening commented on lib/galaxy/jobs/runners/kubernetes.py:
```diff
@@ -451,7 +455,8 @@ def check_watched_item(self, job_state):
             log.debug(
-                "Reaching unexpected point for Kubernetes job, where it is not classified as succ., active nor failed.")
+                "Reaching unexpected point for Kubernetes job name %s where it is not classified as succ., active nor failed.", job.name)
+            log.debug("Here's the full job object\n========================================\n%s\n========================================", job.obj)
```
Can we make this debug message a little bit more standard?
Maybe: `log.debug("K8s full job object: (%s)", job.obj)`

ilveroluca replied:
I patched that message @bgruening

ilveroluca committed:
Improve k8s runner debug log message

mvdbeek commented on lib/galaxy/jobs/runners/kubernetes.py:
```diff
@@ -435,10 +435,14 @@ def check_watched_item(self, job_state):
+                    error_file = open(job_state.error_file, 'w')
+                    error_file.write("")
+                    error_file.close()
+                except IOError as e:
```
Do you know where exactly the IOError is being raised ? I'd assume it's in `self.__produce_log_file(job_state)` ?
Or is it because the job directory doesn't exist anymore ?
In either case you can use 
```
with open(job_state.error_file, 'w'):
    pass
```

ilveroluca replied:
It was `self.__produce_log_file(job_state)`:
```
galaxy.jobs.runners ERROR 2018-09-28 16:54:07,723 Unhandled exception checking active jobs
Traceback (most recent call last):
  File "/export/galaxy-central/lib/galaxy/jobs/runners/__init__.py", line 594, in monitor
    self.check_watched_items()
  File "/export/galaxy-central/lib/galaxy/jobs/runners/__init__.py", line 621, in check_watched_items
    new_async_job_state = self.check_watched_item(async_job_state)
  File "/export/galaxy-central/lib/galaxy/jobs/runners/kubernetes.py", line 430, in check_watched_item
    self.__produce_log_file(job_state)
  File "/export/galaxy-central/lib/galaxy/jobs/runners/kubernetes.py", line 540, in __produce_log_file
    logs_file = open(logs_file_path, mode="w")
IOError: [Errno 2] No such file or directory: '/export/galaxy-central/database/job_working_directory/011/11266/galaxy_11266.o'
```

I presume because the job directory doesn't exist anymore.

Regarding your suggestion to use `with open...`, good call:  I'll patch the old code.

ilveroluca committed:
Push k8s log IOError handling into `__produce_log_file`

Handling the IOError within the function should result in other call
sites within the kubernets.py module properly setting the job state even
if the job directory has been removed.

ilveroluca committed:
Prefer `with open` syntax within kubernetes.py

pcm32 commented on lib/galaxy/jobs/runners/kubernetes.py:
```diff
@@ -543,23 +535,27 @@ def fail_job(self, job_state):
-                log.info("Could not write pod\'s " + pod_obj['metadata']['name'] +
-                         " log file due to HTTPError " + str(detail))
+                log.info("Could not write log file for pod %s due to HTTPError %s",
+                         detail, pod_obj['metadata']['name'])
```
I think that you have objects order swapped here, pod_obj['metadata']['name'] should go first if I'm not mistaken.

ilveroluca replied:
Thank you  @pcm32 :bowing_man:

ilveroluca committed:
Fix inverted log message arguments

Thanks @pcm32!

mvdbeek commented on lib/galaxy/jobs/runners/kubernetes.py:
```diff
@@ -533,23 +535,27 @@ def fail_job(self, job_state):
-                         " log file due to HTTPError " + str(detail))
+                log.info("Could not write log file for pod %s due to HTTPError %s",
+                         pod_obj['metadata']['name'], detail)
+        if isinstance(log_string, text_type):
```
This isn't necessary, a file opened in text mode in python 2 can write bytestring and unicode strings. However as it is now it will fail on python 3, since you get a bytestring after encoding, while you need a unicode string.

mvdbeek replied:
I can do this in a followup.

ilveroluca replied:
Thanks @mvdbeek.  Sorry about the delay in my follow-up.

dannon commented on lib/galaxy/jobs/runners/kubernetes.py:
```diff
@@ -515,8 +512,13 @@ def fail_job(self, job_state):
+        except IOError as e:
+            stdout_content = "<Failed to recuperate log for job {} >".format(job_state.job_id)
+            log.error("Failed to get stdout for job %s", job_state.job_id)
+            log.exception(e)
```
I'd consolidate these two to `log.exception("Failed to get stdout for job %s", job_state.job_id)` and drop 'as e' above.

dannon commented on lib/galaxy/jobs/runners/kubernetes.py:
```diff
@@ -533,23 +535,27 @@ def fail_job(self, job_state):
+                logs_file.write(log_string)
+        except IOError as e:
+            log.error("Couldn't produce log files for %s", job_state.job_id)
+            log.exception(e)
```
Same as above, consolidate these to a single `log.exception("Couldn't...")`

dannon reviewed (commented):
Made two notes for @mvdbeek's followup, but this is good, thanks!  +1
